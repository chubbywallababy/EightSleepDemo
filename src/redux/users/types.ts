export const FETCH_USERS_REQUEST = 'FETCH_USERS_REQUEST';
export const FETCH_USERS_SUCCESS = 'FETCH_USERS_SUCCESS';
export const FETCH_USERS_FAILURE = 'FETCH_USERS_FAILURE';

// Define type for API status
export enum ApiStatus {
  IDLE = 'idle',
  LOADING = 'loading',
  FAILED = 'failed',
}
