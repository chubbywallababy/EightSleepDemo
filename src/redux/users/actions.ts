import {createAction} from '@reduxjs/toolkit';
import {User} from '../../types';

const USER_PREFIX = 'user/';

export const usersActions = {
  fetchUsers: createAction(USER_PREFIX + 'fetch'),
  setUsers: createAction<User[]>(USER_PREFIX + 'set'),
  setError: createAction<string>(USER_PREFIX + 'error'),
};
