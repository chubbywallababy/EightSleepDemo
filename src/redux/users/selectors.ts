import {createSelector} from '@reduxjs/toolkit';
import {RootState} from '../store';

export const selectUsersState = (state: RootState) => state.users;

export const selectUsers = createSelector(
  selectUsersState,
  usersState => usersState.data.users,
);

export const selectHasUserMadeSelection = (id: string) =>
  createSelector(
    selectUsersState,
    usersState => typeof usersState.data.acceptedSuggestion[id] === 'boolean',
  );

export const didUserAcceptSelection = (id: string) =>
  createSelector(
    selectUsersState,
    usersState => usersState.data.acceptedSuggestion[id],
  );

export const selectUsersStatus = createSelector(
  selectUsersState,
  usersState => usersState.status,
);

export const selectUsersError = createSelector(
  selectUsersState,
  usersState => usersState.error,
);
