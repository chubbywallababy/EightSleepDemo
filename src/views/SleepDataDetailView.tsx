import React, {useCallback, useEffect, useState} from 'react';
import {ScrollView, StyleSheet, Text, View} from 'react-native';
import {NativeStackScreenProps} from '@react-navigation/native-stack';
import {RootStackParamList} from '../../App';
import {
  getHeartRates,
  getMovements,
  getRespirations,
  getSortedStages,
} from '../utils/SleepDataUtils';
import {SleepDataByStageChart} from '../components/SleepDataByStageChart';
import {getLineData} from '../components/utils/getLineData';
import {SleepStageSwitch} from '../components/SleepStageSwitch';
import {Stage} from '../types';
import {SleepDataByLineItem} from '../types/SleepDataByLineItem';
import {SleepStagesPieGraph} from '../components/SleepStagesPieGraph';

type SleepDataDetailViewProps = NativeStackScreenProps<
  RootStackParamList,
  'Details'
>;

export const SleepDataDetailView = ({
  route,
  navigation,
}: SleepDataDetailViewProps) => {
  const [selectedStage, setSelectedStage] = useState<Stage>(Stage.Deep);

  const data = route.params.data;

  useEffect(() => {
    navigation?.setOptions({title: route.params.data.date});
  }, [navigation, route.params.data.date]);

  const getCurrentField = useCallback(
    (givenData: SleepDataByLineItem) => {
      switch (selectedStage) {
        case Stage.Deep:
          return givenData.deep;
        case Stage.Light:
          return givenData.light;
        case Stage.Rem:
          return givenData.rem;
        default:
          throw new Error('Encountered new stage: ' + selectedStage);
      }
    },
    [selectedStage],
  );

  const heartData = getHeartRates(data);
  const moveData = getMovements(data);
  const respData = getRespirations(data);

  const tempData = getLineData(getSortedStages(data), 'temperature');

  return (
    <View>
      <View style={styles.switchContainer}>
        <SleepStageSwitch stage={selectedStage} setStage={setSelectedStage} />
      </View>
      <ScrollView>
        <SleepStagesPieGraph />
        <SleepDataByStageChart
          data={getCurrentField(tempData)}
          title="Temperature"
          units={'Celcius'}
          yAxisOffset={30}
          spacing={22}
          noOfSections={3}
          initialSpacing={10}
        />
        <Text style={styles.textField}>{JSON.stringify(tempData)}</Text>
        <Text style={styles.textField}>{JSON.stringify(heartData)}</Text>
        <Text style={styles.textField}>{JSON.stringify(moveData)}</Text>
        <Text style={styles.textField}>{JSON.stringify(respData)}</Text>
      </ScrollView>
    </View>
  );
};

const styles = StyleSheet.create({
  textField: {
    padding: 20,
  },
  switchContainer: {
    paddingTop: 10,
  },
});
