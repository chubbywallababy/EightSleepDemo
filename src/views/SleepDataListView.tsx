// import React, {useEffect, useState} from 'react';
// import {
//   View,
//   StyleSheet,
//   FlatList,
//   Text,
//   ActivityIndicator,
// } from 'react-native';
// import {useAppDispatch, useAppSelector} from '../redux/hooks';
// import {fetchAsync, selectData, selectStatus} from '../redux/sleeper/sleepSlice';
// import {SleepDataCell} from '../components/SleepDataCell';
// import {useNavigation} from './hooks/useNavigation';
// import {AnimatedNumber} from '../components/AnimatedNumber';
// import {Background} from '../components/common';

// export const SleepDataListView = () => {
//   const navigation = useNavigation();

//   const [didLoad, setDidLoad] = useState(false);
//   const data = useAppSelector(selectData);
//   const status = useAppSelector(selectStatus);

//   const dispatch = useAppDispatch();

//   useEffect(() => {
//     if (!didLoad) {
//       dispatch(fetchAsync());
//       setDidLoad(true);
//     }
//   }, [dispatch, didLoad]);

//   return (
//     <Background style={styles.container}>
//       {status === 'loading' ? <ActivityIndicator size="large" /> : null}
//       {status === 'failed' ? <Text>Sorry, something went wrong...</Text> : null}
//       {data.length === 0 && status === 'idle' ? (
//         <Text>There are no nights available for analysis</Text>
//       ) : null}
//       {data.length > 0 && status === 'idle' ? (
//         <>
//           <View style={styles.heading}>
//             <AnimatedNumber n={data.length} duration={500} />
//             <Text style={styles.headingText}>nights tracked</Text>
//           </View>
//           <FlatList
//             style={styles.list}
//             keyExtractor={item => item.id}
//             data={data}
//             renderItem={({item}) => (
//               <SleepDataCell
//                 data={item}
//                 onPress={() => navigation?.navigate('Details', {data: item})}
//               />
//             )}
//           />
//         </>
//       ) : null}
//     </Background>
//   );
// };

// const styles = StyleSheet.create({
//   container: {
//     flex: 1,
//     width: '100%',
//     justifyContent: 'center',
//     alignItems: 'center',
//   },
//   paddedText: {
//     marginVertical: 100,
//   },
//   heading: {
//     paddingVertical: 20,
//     display: 'flex',
//     flexDirection: 'row',
//     alignItems: 'center',
//   },
//   headingText: {
//     marginLeft: 10,
//     fontSize: 16,
//   },
//   list: {
//     flex: 1,
//     width: '100%',
//   },
// });
