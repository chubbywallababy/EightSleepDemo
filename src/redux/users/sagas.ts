import {call, put} from 'redux-saga/effects';
import {SleepService} from '../../services/SleepService';
import {User} from '../../types';
import {usersActions} from './actions';
import {sleepActions} from '../sleep/actions';

export function* fetchUsers() {
  try {
    const data: User[] = yield call(SleepService.getUsers);
    yield put(usersActions.setUsers(data)); // Dispatch with typed data
    for (const i in data) {
      const id = data[i].id;
      yield put(sleepActions.fetchUserSleep(id));
    }
  } catch (error) {
    /**
     * We would have to coordinate with the BE on error structure.
     */
    yield put(usersActions.setError('Unknown error'));
  }
}
