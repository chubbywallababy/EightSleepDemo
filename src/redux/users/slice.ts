import {createSlice, createAsyncThunk} from '@reduxjs/toolkit';
import {userActions} from './actions';
import {User} from '../../types/User';

// Define async thunk for fetching users
export const fetchUsers = createAsyncThunk<User[], void>(
  userActions.FETCH,
  async () => {
    const response = await fetch(
      'https://s3.amazonaws.com/eight-public/challenge/users.json',
    );
    if (!response.ok) {
      throw new Error('Failed to fetch users');
    }
    const data = await response.json();
    return data.users;
  },
);

// Define initial state for user slice
interface UserState {
  status: 'idle' | 'loading' | 'failed';
  error: string | null;
  data: User[];
}

const initialState: UserState = {
  status: 'idle',
  error: null,
  data: [],
};

// Define user slice
const userSlice = createSlice({
  name: 'user',
  initialState,
  reducers: {},
  extraReducers: builder => {
    builder
      .addCase(fetchUsers.pending, state => {
        state.status = 'loading';
        state.error = null;
      })
      .addCase(fetchUsers.fulfilled, (state, action) => {
        state.status = 'idle';
        state.data = action.payload;
      })
      .addCase(fetchUsers.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.error.message || 'Unknown error';
      });
  },
});

export default userSlice.reducer;
