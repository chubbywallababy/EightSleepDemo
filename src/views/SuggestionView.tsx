import React, {useCallback} from 'react';
import {TouchableOpacity, View} from 'react-native';
import {CardTitle, SleepText} from '../components/common';
import {strings} from '../i18n';
import {NativeStackScreenProps} from '@react-navigation/native-stack';
import {RootStackParamList} from '../../App';
import {StyleSheet} from 'react-native';
import {colors} from '../styles/colors';
import {useAppDispatch} from '../redux/hooks';
import {acceptSuggestion, denySuggestion} from '../redux/users/slice';

type SuggestionViewProps = NativeStackScreenProps<
  RootStackParamList,
  'Suggestion'
>;

/**
 * A view for a user to view and accept suggestions. This is currently only a view with text and buttons.
 *
 * Future improvements would be to include a graph like the one in the mocks
 */
export const SuggestionView = ({route, navigation}: SuggestionViewProps) => {
  const dispatch = useAppDispatch();

  const makeSelection = useCallback(
    (accept: boolean) => {
      if (accept) {
        dispatch(acceptSuggestion(route.params.data.id));
      } else {
        dispatch(denySuggestion(route.params.data.id));
      }
      navigation.goBack();
    },
    [dispatch, route, navigation],
  );

  return (
    <View style={styles.container}>
      <View style={styles.cardContainer}>
        <CardTitle style={[styles.text, styles.title]}>
          {strings.suggestion.title}
        </CardTitle>
        <SleepText style={[styles.text, styles.suggestionText]}>
          {strings.suggestion.text}
        </SleepText>
      </View>
      <View style={styles.buttonContainer}>
        <TouchableOpacity
          onPress={() => makeSelection(true)}
          style={[styles.button, styles.primaryButton]}>
          <SleepText style={styles.buttonText}>
            {strings.common.acceptChanges}
          </SleepText>
        </TouchableOpacity>
        <TouchableOpacity
          onPress={() => makeSelection(false)}
          style={styles.button}>
          <SleepText style={styles.buttonText}>
            {strings.common.notTonight}
          </SleepText>
        </TouchableOpacity>
      </View>
    </View>
  );
};
const styles = StyleSheet.create({
  container: {
    padding: 8,
    gap: 20,
  },
  cardContainer: {
    padding: 8,
    gap: 12,
  },
  text: {
    textAlign: 'center',
  },
  suggestionText: {
    fontSize: 16,
  },
  title: {
    textTransform: 'capitalize',
  },
  buttonText: {
    fontSize: 14,
    fontWeight: '700',
  },
  button: {
    height: 54,
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: 6,
  },
  primaryButton: {
    backgroundColor: colors.primary,
  },
  buttonContainer: {
    gap: 8,
  },
});
