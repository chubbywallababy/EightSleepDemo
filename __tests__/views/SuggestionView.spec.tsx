import React from 'react';
import {render} from '@testing-library/react-native';
import {Provider} from 'react-redux';
import {mockStore} from '../__mocks__/mockStore';
import {users} from '../__mocks__/users';
import {SuggestionView} from '../../src/views/SuggestionView';
import {TouchableOpacity} from 'react-native';

describe('boilerplate for SuggestionView', () => {
  it('should render the view', async () => {
    // Render the app
    const {root} = render(
      <Provider store={mockStore}>
        <SuggestionView
          route={{params: {data: users[1]}, key: '', name: 'Suggestion'}}
          navigation={
            {
              navigate: jest.fn(),
              dispatch: jest.fn(),
              reset: jest.fn(),
              goBack: jest.fn(),
              isFocused: jest.fn(),
              canGoBack: jest.fn(),
              getId: jest.fn(),
              getState: jest.fn(),
              getParent: jest.fn(),
              setOptions: jest.fn(),
              setParams: jest.fn(),
              addListener: jest.fn(),
              removeListener: jest.fn(),
            } as any
          }
        />
      </Provider>,
    );

    const view = root.findAllByType(TouchableOpacity);
    expect(view).toHaveLength(2);
  });
});
