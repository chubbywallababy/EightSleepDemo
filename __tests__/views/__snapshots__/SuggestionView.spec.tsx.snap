// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`boilerplate for SuggestionView should render the view 1`] = `
<View
  style={
    {
      "gap": 20,
      "padding": 8,
    }
  }
>
  <View
    style={
      {
        "gap": 12,
        "padding": 8,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "white",
          },
          [
            {
              "color": "white",
            },
            {
              "fontSize": 22,
              "fontWeight": "700",
            },
            [
              {
                "textAlign": "center",
              },
              {
                "textTransform": "capitalize",
              },
            ],
          ],
        ]
      }
    >
      Temperature suggestion
    </Text>
    <Text
      style={
        [
          {
            "color": "white",
          },
          [
            {
              "textAlign": "center",
            },
            {
              "fontSize": 16,
            },
          ],
        ]
      }
    >
      Based on your recent activity, we suggest a cooler temperature for your Pod tonight.
    </Text>
  </View>
  <View
    style={
      {
        "gap": 8,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#246AFF",
          "borderRadius": 6,
          "height": 54,
          "justifyContent": "center",
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "white",
            },
            {
              "fontSize": 14,
              "fontWeight": "700",
            },
          ]
        }
      >
        Accept changes
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "borderRadius": 6,
          "height": 54,
          "justifyContent": "center",
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "white",
            },
            {
              "fontSize": 14,
              "fontWeight": "700",
            },
          ]
        }
      >
        Not tonight
      </Text>
    </View>
  </View>
</View>
`;
